idx/memory/tag_query.go

```diff
@@ -0,0 +1,176 @@
+package memory
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/raintank/metrictank/idx"
+)
+
+var (
+	expressionRe    = regexp.MustCompile("^([^!=~]+)([!=~]{1,3})(.*)$")
+	errUnknownId    = errors.New("id does not exist")
+	errInvalidTag   = errors.New("found an invalid tag in the index")
+	errInvalidQuery = errors.New("invalid query")
+)
+
+type TagQuery struct {
+	selects []func(TagIndex) (map[string]struct{}, error)
+	filters []func(def *idx.Archive) (bool, error)
+}
+
+func NewTagQuery(expressions []string) (*TagQuery, error) {
+	if len(expressions) == 0 {
+		return nil, errInvalidQuery
+	}
+
+	query := &TagQuery{}
+
+	for _, expr := range expressions {
+		if !expressionRe.MatchString(expr) {
+			return nil, errInvalidQuery
+		}
+
+		exprSplits := expressionRe.FindAllStringSubmatch(expr, -1)
+		if len(exprSplits[0]) != 4 {
+			return nil, errInvalidQuery
+		}
+
+		key := exprSplits[0][1]
+		operator := exprSplits[0][2]
+		value := exprSplits[0][3]
+
+		// special case of empty value
+		if len(value) == 0 {
+			if operator[0] == byte('=') {
+				operator = "!=~"
+				value = ".+"
+			} else {
+				operator = "=~"
+				value = ".+"
+			}
+		}
+
```

Comment:
I think we can do better in handling terms that match the empty string, since we have a requirement that there must be a tag expression that doesn't match the empty string.  Because of that requirement, we should be able to make a list of candidate series from the expression that requires a match, then use those as the list of "all" series that should be considered to match if they don't have the tags that we're matching the empty string for.  I didn't implement that concept in the graphite tagdbs, but I'm going to go back and see whether I can use it.

Reply:
so i'm not completely sure i understand you right. but actually i think this already does what you describe:

- further down queries are divided into "selects" and "filters". any query that's excluding something is a "filter", any query that's selecting by a match is a "select".
- a query like `key=` is translated into `key!=.+`, this is a filter
- the "select" queries are always executed first, building that list of candidates you mention
- after the list of candidates is built the "filters" are applied

we can further speed this up by not even matching the regex `.+` because we can't have empty tag values in the index. so every series that has they tag `key` will match the value `.+`, so we can skip the regex match.